pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -1171,34 +1186,50 @@ func (c *Reconciler) updatePipelineRunStatusFromInformer(ctx context.Context, pr
 		logger.Errorf("could not list TaskRuns %#v", err)
 		return err
 	}
-	updatePipelineRunStatusFromTaskRuns(logger, pr, taskRuns)
+	updatePipelineRunStatusFromTaskRuns(ctx, logger, pr, taskRuns)
 
 	runs, err := c.runLister.Runs(pr.Namespace).List(k8slabels.SelectorFromSet(pipelineRunLabels))
 	if err != nil {
 		logger.Errorf("could not list Runs %#v", err)
 		return err
 	}
-	updatePipelineRunStatusFromRuns(logger, pr, runs)
+	updatePipelineRunStatusFromRuns(ctx, logger, pr, runs)
 
 	return nil
 }
 
-func updatePipelineRunStatusFromTaskRuns(logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, trs []*v1beta1.TaskRun) {
+func updatePipelineRunStatusFromTaskRuns(ctx context.Context, logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, trs []*v1beta1.TaskRun) {
 	// If no TaskRun was found, nothing to be done. We never remove taskruns from the status
 	if trs == nil || len(trs) == 0 {
 		return
 	}
+
+	cfg := config.FromContextOrDefaults(ctx)
+	fullEmbedded := cfg.FeatureFlags.EmbeddedStatus == config.FullEmbeddedStatus || cfg.FeatureFlags.EmbeddedStatus == config.BothEmbeddedStatus
+	minimalEmbedded := cfg.FeatureFlags.EmbeddedStatus == config.MinimalEmbeddedStatus || cfg.FeatureFlags.EmbeddedStatus == config.BothEmbeddedStatus
+
 	// Store a list of Condition TaskRuns for each PipelineTask (by name)
 	conditionTaskRuns := make(map[string][]*v1beta1.TaskRun)
 	// Map PipelineTask names to TaskRun names that were already in the status
 	taskRunByPipelineTask := make(map[string]string)
-	if pr.Status.TaskRuns != nil {
-		for taskRunName, pipelineRunTaskRunStatus := range pr.Status.TaskRuns {
-			taskRunByPipelineTask[pipelineRunTaskRunStatus.PipelineTaskName] = taskRunName
+	// Map PipelineTask names to TaskRun child references that were already in the status
+	childRefByPipelineTask := make(map[string]*v1beta1.ChildStatusReference)
+
+	if fullEmbedded {
+		if pr.Status.TaskRuns != nil {
+			for taskRunName, pipelineRunTaskRunStatus := range pr.Status.TaskRuns {
+				taskRunByPipelineTask[pipelineRunTaskRunStatus.PipelineTaskName] = taskRunName
+			}
+		} else {
+			pr.Status.TaskRuns = make(map[string]*v1beta1.PipelineRunTaskRunStatus)
+		}
+	}
+	if minimalEmbedded {
+		for i := range pr.Status.ChildReferences {
+			childRefByPipelineTask[pr.Status.ChildReferences[i].PipelineTaskName] = &pr.Status.ChildReferences[i]
```

Comment:
Yup, you're right - good catch! Thanks! =)

Reply:
In that case, maybe it is better to create a new function `updatePipelineRunStatusFromChildRefs` and call that function when the feature flag is set to "minimal" or "both".

Reply:
I'm pushing an update momentarily, but plan to come back to this later this afternoon and see if I can come up with a good way to rewrite the whole `updatePipelineRunStatusFromInformer` to make more sense with `ChildReferences` involved now.

Reply:
So...I'm not sure if there's a better way to rework this (beyond removing a return I had in at one point but stopped using, which I'm pushing momentarily) - it's going to be messy no matter what, because we're doing very similar things for both the `full` and `minimal` cases, and for `both`, we obviously have to do both. I think I prefer splitting it up into "do the stuff relating to `TaskRun`s" and "do the stuff relating to `Run`s", rather than "do the stuff for `full`/`both`" and "do the stuff for `minimal`/`both`". But I could be wrong, so I'm going to dwell on this a bit longer.

Reply:
I made a run at it, but it just ended up messier than what's already there. No matter what we do, this logic is going to be messy.

Reply:
Ok, I got hooked on trying to figure out a less messy way and am maybe making some progress, but I forgot another thing that makes it more complicated (besides what I mentioned in https://github.com/tektoncd/pipeline/pull/4706#discussion_r839942364) - we have to generate the `TaskRun`/`Run` name for new ones, and if we're handling updating `TaskRuns`/`Runs` separately from updating `ChildReferences`, then when we've got `embedded-status='both'`, we need to make sure that we only generate a new name once. So, with `ChildReferences` as our first choice for source of truth, when we go to update `TaskRuns` after updating `ChildReferences`, we have to use the new `TaskRun` name we generated when updating `ChildReferences` again in `TaskRuns`.

I've made real progress in significantly simplifying the condition check handling part, so if I can nail the right way to do the `TaskRun` name generation once-and-only-once, I may be able to refactor this. It'll still be fairly ugly and have a good amount of duplicated code, but I think the current mess is the only way to avoid the duplicated code.

Reply:
@lbernick Ok! I think I may have a better approach - see https://github.com/tektoncd/pipeline/pull/4706/commits/18521d4dd6e5802d56e6687bcc8ad4c2b6071ac0. It's not perfect yet - needs some cleanup/comments/test changes (specifically tests for the new `updatePipelineRunStatusFromSlices`, `getNewConditionChecksForTaskRun`,  and `updatePipelineRunStatusFromChildRefs`, a test for `updatePipelineRunStatusFromRuns`, and changes to the test for `updatePipelineRunStatusFromTaskRuns`), but it does decouple and simplify things. I'd love your thoughts!